src/content/tutorials/user-auth-with-auth0/_index.md

```diff
@@ -0,0 +1,736 @@
+  }
+
+  return { headers, status: 302 }
+}
+```
+
+With our authorization logic defined, we can finally wrap up the corresponding code in `workers-site/index.js`. Knowing that the `handleRedirect` function will pass back an object with `Response` options, we can make a new `Response` object, passing in the information from the existing `response`, and adding the new headers and status code from `handleRedirect`. This will redirect the user to the app's root path, setting a cookie to indicate that they're authorized for future requests:
+
+```js
+// workers-site/index.js
+
+async function handleEvent(event) {
+  try {
+    // BEGINNING OF HANDLE AUTH REDIRECT CODE BLOCK
+    if (url.pathname === '/auth') {
+      const authorizedResponse = await handleRedirect(event)
+      response = new Response(response.body, {
+        response,
+        ...authorizedResponse,
+      })
+      return response
+    }
+    // END OF HANDLE AUTH REDIRECT CODE BLOCK
+
+    // BEGINNING OF WORKERS SITES
+  }
+}
+```
+
+## Adding the verify read block
+
+With our application persisting authentication data in KV, and associating it to the current user via a cookie, we're now prepared to fill out the `verify` function defined earlier in the tutorial. This function will look at the `Cookie` header, and look up the authentication info that we persisted in Workers KV. The tokens saved in Workers KV can be used to make a subrequest to Auth0, which has the benefit of both re-verifying the user, as well as getting some user info, such as name and email address, if they're provided using the user's authentication method of choice.
+
+To begin, install the NPM package [`cookie`](https://www.npmjs.com/package/cookie), which we'll use to simplify parsing the `Cookie` header in the `request`:
+
+```sh
+$ npm install cookie
+```
+
+In `workers-site/auth0.js`, we can begin to flesh out the contents of the `verify` function, beginning by parsing the `Cookie` header, looking for our `cookieKey` as defined earlier in the tutorial:
+
+```js
+// workers-site/auth0.js
+
+import cookie from 'cookie'
+
+const verify = async event => {
+  const cookieHeader = event.request.headers.get('Cookie')
+  if (cookieHeader && cookieHeader.includes(cookieKey)) {
+    const cookies = cookie.parse(cookieHeader)
+    if (!cookies[cookieKey]) return {}
+    const sub = cookies[cookieKey]
+  }
+  return {}
+}
+```
+
+With the unique ID `sub` parsed from the `Cookie` header, we can use it to retrieve the user information we previously stored in KV:
+
+```js
+// workers-site/auth0.js
+
+const verify = async event => {
+  const cookieHeader = event.request.headers.get('Cookie')
+  if (cookieHeader && cookieHeader.includes(cookieKey)) {
+    const cookies = cookie.parse(cookieHeader)
+    if (!cookies[cookieKey]) return {}
+    const sub = cookies[cookieKey]
+
+    const kvStored = JSON.parse(await AUTH_STORE.get(sub))
```

Comment:
Is it safe to assume `JSON.parse` won't throw here?

Reply:
+1 Users always have the ability to modify their cookie, essentially we're trusting their input and hoping that there's an associated entry in KV. I think spitting back a 4xx or sending an error to console if KV doesn't have an entry associated could be a good idea.